### PR TITLE
Invalid build option warnings - supersedes #1088

### PR DESCRIPTION
**Pull request by MatzFan**
- [ √] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [√ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
  Note this PR supersedes #1088
- [ √] Have you added an explanation of what your changes do and why you'd like us to include them?
- [√ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [√ ] Have you successfully run `brew tests` with your changes locally?
  Be grateful if someone could explain how to do this in a forked repo outside of /usr/local/Homebrew 😄 

---

### CODE REVIEW

File: Library/Homebrew/build_options.rb [hunk 1]
```diff
@@ -101,6 +101,15 @@ def unused_options
     @options - @args
   end
 
+  # @private
+  def invalid_options
+    @args - @options
+  end
+
+  def invalid_option_names
```
Comment (MikeMcQuaid) [inline Library/Homebrew/build_options.rb]: Make this `@private` too, thanks!

File: Library/Homebrew/formula_installer.rb [hunk 2]
```diff
@@ -32,7 +32,7 @@ def self.mode_attr_accessor(*names)
   end
 
   attr_reader :formula
-  attr_accessor :options, :build_bottle
+  attr_accessor :options, :build_bottle, :invalid_opt_names
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb]: Let's call this `invalid_option_names`

File: Library/Homebrew/formula_installer.rb [hunk 3]
```diff
@@ -214,6 +215,10 @@ def install
       opoo "#{formula.full_name}: #{old_flag} was deprecated; using #{new_flag} instead!"
     end
 
+    invalid_opt_names.each do |option|
+      opoo "#{formula.full_name}: #{option} is invalid for this formula and will be ignored!"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb]: `opoo "#{formula.full_name}: this formula has no #{option} option so it will be ignored!"`

**Comment by MikeMcQuaid** [comment]
Just a heads up that https://github.com/Homebrew/brew/pull/1488 may affect this PR. Let me know your thoughts.

File: Library/Homebrew/cmd/install.rb [hunk 4]
```diff
@@ -261,7 +261,9 @@ def install_formula(f)
     f.print_tap_action
 
     fi = FormulaInstaller.new(f)
-    fi.options             = f.build.used_options
+    build_options          = f.build
+    fi.options             = build_options.used_options
+    fi.invalid_opt_names   = build_options.invalid_option_names
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/install.rb]: `fi.invalid_option_names`

File: Library/Homebrew/cmd/install.rb [hunk 5]
```diff
@@ -261,7 +261,9 @@ def install_formula(f)
     f.print_tap_action
 
     fi = FormulaInstaller.new(f)
-    fi.options             = f.build.used_options
+    build_options          = f.build
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/install.rb]: Pull this above the `fi` definition like in `reinstall.rb` to make it clearer and don't indent it.

Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/install.rb]: Also, wondering if we'll want anything like this in `reinstall` or `upgrade`? I figure probably not but worth thinking about.

**Comment by MatzFan** [comment]
Will have to look at #1488 - can't see right away what it does. Thx for heads up tho. Also, sorry for deleting original branch for this change - a learning experience

**Comment by MikeMcQuaid** [comment]
@MatzFan Actually, I looked and it's ok; you are touching `install`, I'm touching `reinstall`, `upgrade` and dependencies. We're all good 👍

**Comment by MikeMcQuaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

**Pull request merged**